# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 


In [2]:
import pandas as pd
import numpy as np

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [3]:
with open('horse_data.names') as names:
    for i, row in enumerate(names):
        print(i, row)

0 1. TItle: Horse Colic database

1 

2 2. Source Information

3    -- Creators: Mary McLeish & Matt Cecile

4 	  	Department of Computer Science

5 		University of Guelph

6 		Guelph, Ontario, Canada N1G 2W1

7 		mdmcleish@water.waterloo.edu

8    -- Donor:    Will Taylor (taylor@pluto.arc.nasa.gov)

9    -- Date:     8/6/89

10 

11 3. Past Usage:

12    -- Unknown

13 

14 4. Relevant Information:

15 

16    -- 2 data files

17       -- horse-colic.data: 300 training instances

18       -- horse-colic.test: 68 test instances

19    -- Possible class attributes: 24 (whether lesion is surgical)

20      -- others include: 23, 25, 26, and 27

21    -- Many Data types: (continuous, discrete, and nominal)

22 

23 5. Number of Instances: 368 (300 for training, 68 for testing)

24 

25 6. Number of attributes: 28

26 

27 7. Attribute Information:

28 

29   1:  surgery?

30           1 = Yes, it had surgery

31           2 = It was treated without surgery

32 

33   2:  Age

34         

In [4]:
horse_data_csv = pd.read_csv('horse_data.csv', names=['surgery?','Age','Hospital_Number','rectal_temperature','pulse',
                                                  'respiratory_rate', 'temperature_of_extremities', 'peripheral_pulse',
                                                  'mucous_membranes', 'capillary_refill_time', 'pain', 'peristalsis',
                                                  'abdominal_distension', 'nasogastric_tube', 'nasogastric_reflux',
                                                  'nasogastric_reflux_PH', 'rectal_examination-feces', 'abdomen',
                                                  'packed_cell_volume', 'total_protein', 'abdominocentesis_appearance',
                                                  'abdomcentesis_total_protein', 'outcome', 'surgical_lesion?',
                                                  'type_of_lesion', '26', '27', 'cp_data'
                                                 ])
horse_data_csv

surgery?  Age  Hospital_Number rectal_temperature pulse respiratory_rate  \
0          2    1           530101              38.50    66               28   
1          1    1           534817               39.2    88               20   
2          2    1           530334              38.30    40               24   
3          1    9          5290409              39.10   164               84   
4          2    1           530255              37.30   104               35   
..       ...  ...              ...                ...   ...              ...   
295        1    1           533886                  ?   120               70   
296        2    1           527702              37.20    72               24   
297        1    1           529386              37.50    72               30   
298        1    1           530612              36.50   100               24   
299        1    1           534618               37.2    40               20   

    temperature_of_extremities peripheral_pulse mucous_membranes  \
0                            3                3                ?   
1                            ?                ?                4   
2                            1                1                3   
3                            4                1                6   
4                            ?                ?                6   
..                         ...              ...              ...   
295                          4                ?                4   
296                          3                2                4   
297                          4                3                4   
298                          3                3                3   
299                          ?                ?                ?   

    capillary_refill_time  ... packed_cell_volume total_protein  \
0                       2  ...              45.00          8.40   
1                       1  ...                 50            85   
2                       1  ...              33.00          6.70   
3                       2  ...              48.00          7.20   
4                       2  ...              74.00          7.40   
..                    ...  ...                ...           ...   
295                     2  ...                 55            65   
296                     2  ...              44.00             ?   
297                     1  ...              60.00          6.80   
298                     1  ...              50.00          6.00   
299                     ?  ...                 36            62   

    abdominocentesis_appearance abdomcentesis_total_protein outcome  \
0                             ?                           ?       2   
1                             2                           2       3   
2                             ?                           ?       1   
3                             3                        5.30       2   
4                             ?                           ?       2   
..                          ...                         ...     ...   
295                           ?                           ?       3   
296                           3                        3.30       3   
297                           ?                           ?       2   
298                           3                        3.40       1   
299                           1                           1       3   

    surgical_lesion? type_of_lesion 26 27 cp_data  
0                  2          11300  0  0       2  
1                  2           2208  0  0       2  
2                  2              0  0  0       1  
3                  1           2208  0  0       1  
4                  2           4300  0  0       2  
..               ...            ... .. ..     ...  
295                2           3205  0  0       2  
296                1           2208  0  0       1  
297                1           3205  0  0       2  
298                1           2208  0  0       

In [10]:
horse_data_csv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   surgery?                     300 non-null    object
 1   Age                          300 non-null    int64 
 2   Hospital_Number              300 non-null    int64 
 3   rectal_temperature           300 non-null    object
 4   pulse                        300 non-null    object
 5   respiratory_rate             300 non-null    object
 6   temperature_of_extremities   300 non-null    object
 7   peripheral_pulse             300 non-null    object
 8   mucous_membranes             300 non-null    object
 9   capillary_refill_time        300 non-null    object
 10  pain                         300 non-null    object
 11  peristalsis                  300 non-null    object
 12  abdominal_distension         300 non-null    object
 13  nasogastric_tube             300 no

In [105]:
horse_data = horse_data_csv.mask(horse_data_csv == '?')
horse_data = horse_data.astype({'Age' : 'object',
                                'Hospital_Number' : 'object',
                                'rectal_temperature' : 'float',
                                'pulse' : 'float',
                                'respiratory_rate' : 'float',
                                'temperature_of_extremities' : 'float',
                                'peripheral_pulse' : 'float',
                                'mucous_membranes' : 'float',
                                'nasogastric_reflux_PH' : 'float',
                                'packed_cell_volume' : 'float',
                                'total_protein' : 'float',
                                'abdomcentesis_total_protein' : 'float',
                                'surgical_lesion?' : 'object',
                                'packed_cell_volume' : 'object',
                                'type_of_lesion' : 'object',
                                '26' : 'object',
                                '27' : 'object',
                                'cp_data' : 'object'
                                })
horse_data

surgery? Age Hospital_Number  rectal_temperature  pulse  respiratory_rate  \
0          2   1          530101                38.5   66.0              28.0   
1          1   1          534817                39.2   88.0              20.0   
2          2   1          530334                38.3   40.0              24.0   
3          1   9         5290409                39.1  164.0              84.0   
4          2   1          530255                37.3  104.0              35.0   
..       ...  ..             ...                 ...    ...               ...   
295        1   1          533886                 NaN  120.0              70.0   
296        2   1          527702                37.2   72.0              24.0   
297        1   1          529386                37.5   72.0              30.0   
298        1   1          530612                36.5  100.0              24.0   
299        1   1          534618                37.2   40.0              20.0   

     temperature_of_extremities  peripheral_pulse  mucous_membranes  \
0                           3.0               3.0               NaN   
1                           NaN               NaN               4.0   
2                           1.0               1.0               3.0   
3                           4.0               1.0               6.0   
4                           NaN               NaN               6.0   
..                          ...               ...               ...   
295                         4.0               NaN               4.0   
296                         3.0               2.0               4.0   
297                         4.0               3.0               4.0   
298                         3.0               3.0               3.0   
299                         NaN               NaN               NaN   

    capillary_refill_time  ... packed_cell_volume total_protein  \
0                       2  ...              45.00           8.4   
1                       1  ...                 50          85.0   
2                       1  ...              33.00           6.7   
3                       2  ...              48.00           7.2   
4                       2  ...              74.00           7.4   
..                    ...  ...                ...           ...   
295                     2  ...                 55          65.0   
296                     2  ...              44.00           NaN   
297                     1  ...              60.00           6.8   
298                     1  ...              50.00           6.0   
299                   NaN  ...                 36          62.0   

    abdominocentesis_appearance abdomcentesis_total_protein outcome  \
0                           NaN                         NaN       2   
1                             2                         2.0       3   
2                           NaN                         NaN       1   
3                             3                         5.3       2   
4                           NaN                         NaN       2   
..                          ...                         ...     ...   
295                         NaN                         NaN       3   
296                           3                         3.3       3   
297                         NaN                         NaN       2   
298                           3                         3.4       1   
299                           1                         1.0       3   

     surgical_lesion? type_of_lesion 26 27  cp_data  
0                   2          11300  0  0        2  
1                   2           2208  0  0        2  
2                   2              0  0  0        1  
3                   1           2208  0  0        1  
4                   2           4300  0  0        2  
..                ...            ... .. ..      ...  
295                 2           3205  0  0        2  
296                 1           2208  0  0        1  
297                 1           32

In [106]:
horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   surgery?                     299 non-null    object 
 1   Age                          300 non-null    object 
 2   Hospital_Number              300 non-null    object 
 3   rectal_temperature           240 non-null    float64
 4   pulse                        276 non-null    float64
 5   respiratory_rate             242 non-null    float64
 6   temperature_of_extremities   244 non-null    float64
 7   peripheral_pulse             231 non-null    float64
 8   mucous_membranes             253 non-null    float64
 9   capillary_refill_time        268 non-null    object 
 10  pain                         245 non-null    object 
 11  peristalsis                  256 non-null    object 
 12  abdominal_distension         244 non-null    object 
 13  nasogastric_tube    

In [120]:
my_df = horse_data[['surgery?', 'Age', 'Hospital_Number', 'rectal_temperature', 'pulse', 'pain', 'total_protein', 'outcome']]
my_df.describe(include = 'all')

surgery?    Age  Hospital_Number  rectal_temperature       pulse pain  \
count       299  300.0            300.0          240.000000  276.000000  245   
unique        2    2.0            284.0                 NaN         NaN    5   
top           1    1.0         530693.0                 NaN         NaN    3   
freq        180  276.0              2.0                 NaN         NaN   67   
mean        NaN    NaN              NaN           38.167917   71.913043  NaN   
std         NaN    NaN              NaN            0.732289   28.630557  NaN   
min         NaN    NaN              NaN           35.400000   30.000000  NaN   
25%         NaN    NaN              NaN           37.800000   48.000000  NaN   
50%         NaN    NaN              NaN           38.200000   64.000000  NaN   
75%         NaN    NaN              NaN           38.500000   88.000000  NaN   
max         NaN    NaN              NaN           40.800000  184.000000  NaN   

        total_protein outcome  
count      267.000000     299  
unique            NaN       3  
top               NaN       1  
freq              NaN     178  
mean        24.456929     NaN  
std         27.475009     NaN  
min          3.300000     NaN  
25%          6.500000     NaN  
50%          7.500000     NaN  
75%         57.000000     NaN  
max         89.000000     NaN

In [123]:
Hospital_Number = horse_data['Hospital_Number'].apply(lambda x: len(str(x)))
Hospital_Number.value_counts()

6    265
7     35
Name: Hospital_Number, dtype: int64

In [126]:
len( horse_data [horse_data['pulse'] <= 40 ] )

24

In [127]:
print(horse_data['pain'].describe(),'\n')
print(horse_data['pain'].unique(),'\n')
print(horse_data['pain'].value_counts(),'\n')

count     245
unique      5
top         3
freq       67
Name: pain, dtype: object 

['5' '3' '2' nan '4' '1'] 

3    67
2    59
5    42
4    39
1    38
Name: pain, dtype: int64 



In [128]:
print(horse_data['outcome'].describe(),'\n')
print(horse_data['outcome'].unique(),'\n')
print(horse_data['outcome'].value_counts(),'\n')

count     299
unique      3
top         1
freq      178
Name: outcome, dtype: object 

['2' '3' '1' nan] 

1    178
2     77
3     44
Name: outcome, dtype: int64 



### Ответ

1. surgery?:
    * Как минимум 180 из 300 лошадей было прооперированно.
    * Данные почти полностью присутствуют, выбросов нет.
2. Age:
    * 276 из 300 лощадей взрослые. Остальные 24 - моложе 6 месяцев.
    * Данные присутствуют полностью, выбросов нет.
3. Hospital_Number:
    * На 300 лошадей приходится 284 уникальных дела. При этом каждое не уникальное дело встречается не больше 2х раз.
      ===> 8 лощадей прошли лечение 2 раза, остальные - по одному.
    * 265 номеров дел - шестизначные, остальные - семизначные.
    * Данные полные, выбросов нет.
4. rectal_temperature
    * Средняя температура по больнице на 0,3 градуса выше нормы xD
    * Данные присутствуют не полностью (20% пропусков). Также возможно есть выбросы, связанные с погрешностью измерения или чем-либо еще.
    * Проценталь 25% находится на нормальной температуре 37,8 градусов (у 17 лошадей такая темперутура). Что означает, что у большинства лошадей температура повышена.
5. pulse
    * В данных присутствуют незначительные пропуски (8%).
    * У абсолютного большинства лошадей пульс повышенный (только 24 лошади имеют пульс 40 и ниже).
    * Есть выбросы с очень выскоим значением.
6. pain
    * Присутствует 19% пропусков. Выбросов нет.
    * У большинства (67) лошадей подавленное состояние. 59 испытывают периодическую лёгкую боль. Остальные примерно поровну испытывают тревогу (38) / периодическую сильную боль (39) / постоянную сильную боль (42).
7. total_protein
    * 11% пропусков. Возможны выбросы.
    * Медиана расположена на значении 7,5 - верхней границе допустимой нормы. Следовательно почти половина лошадей испытывали обезвожевание.
8. outcome
    * Данные представлены почти в идеальном виде: нет выбросов, всего один пропуск.
    * 178 лошадей выжило, 77 умерло, 44 были усыплены.
    * (Вопрос: Как лучше считать количество каждого уникального элемента в датафрейме / сериес?)

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [145]:
new_df = my_df
new_df.describe(include = 'all')

surgery?    Age  Hospital_Number  rectal_temperature       pulse pain  \
count       299  300.0            300.0          240.000000  276.000000  245   
unique        2    2.0            284.0                 NaN         NaN    5   
top           1    1.0         530693.0                 NaN         NaN    3   
freq        180  276.0              2.0                 NaN         NaN   67   
mean        NaN    NaN              NaN           38.167917   71.913043  NaN   
std         NaN    NaN              NaN            0.732289   28.630557  NaN   
min         NaN    NaN              NaN           35.400000   30.000000  NaN   
25%         NaN    NaN              NaN           37.800000   48.000000  NaN   
50%         NaN    NaN              NaN           38.200000   64.000000  NaN   
75%         NaN    NaN              NaN           38.500000   88.000000  NaN   
max         NaN    NaN              NaN           40.800000  184.000000  NaN   

        total_protein outcome  
count      267.000000     299  
unique            NaN       3  
top               NaN       1  
freq              NaN     178  
mean        24.456929     NaN  
std         27.475009     NaN  
min          3.300000     NaN  
25%          6.500000     NaN  
50%          7.500000     NaN  
75%         57.000000     NaN  
max         89.000000     NaN

Выбросы присутствуют в стобцах:
* rectal_temperature
* pulse
* total_protein  


1. rectal_temperature:  
    Тут выбросы скорее всего связаны с погрешностью измерений. Можно попробовать найти выбросы через межквартальный размах и отсечь слишком высокие / слишком низкий результаты.
2. pulse:  
    Присутствуют очень высокие значения. Возможно человеческий фактор при записи / измерении, либо же какие-то аномальные значения пульса, которые в любом случае в статистике ни к чему. Делаем то же, что в п.1.
3. total_protein:  
    То же, что в п.2.

In [147]:
def find_outlier(df, key):
    q1 = df[key].quantile(0.25)
    q3 = df[key].quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr) 
    upper_bound = q3 + (1.5 * iqr)
    remove_outliers = df[df[key].between(lower_bound, upper_bound, inclusive=True)].sort_values(str(key))
    return remove_outliers

In [148]:
new_df = find_outlier(new_df, 'rectal_temperature')
new_df

surgery? Age Hospital_Number  rectal_temperature  pulse pain  \
113        1   1          527933                36.8   60.0  NaN   
277        2   1          528620                36.9   50.0    1   
70         2   1          528214                37.0   60.0    3   
292        1   1          530034                37.0   66.0    4   
190        1   1          528742                37.1   40.0    3   
..       ...  ..             ...                 ...    ...  ...   
165        1   1          534925                39.4  120.0  NaN   
117        2   1         5290482                39.5    NaN    3   
231        2   1          530028                39.5   92.0    5   
287        2   9         5305129                39.5   84.0  NaN   
162        2   1          535176                39.5   60.0    3   

     total_protein outcome  
113            NaN       2  
277            6.5       1  
70             7.6       3  
292            6.9       2  
190            6.7       1  
..             ...     ...  
165           64.0       3  
117            6.7       3  
231            6.4       2  
287            5.0       1  
162           56.0       1  

[226 rows x 8 columns]

In [149]:
new_df = find_outlier(new_df, 'pulse')
new_df

surgery? Age Hospital_Number  rectal_temperature  pulse pain  \
232        1   1         5262543                38.5   30.0  NaN   
130        1   1          528964                37.8   30.0  NaN   
276        1   1          534787                38.0   36.0    4   
203        2   1          529685                37.2   36.0    2   
76         1   1          534615                37.6   38.0    3   
..       ...  ..             ...                 ...    ...  ...   
63         2   1          534293                38.2  130.0    2   
23         1   9          534998                38.3  130.0    2   
148        1   1          529272                38.3  132.0    2   
135        1   9         5287179                38.1  136.0    5   
103        1   9         5292489                38.0  140.0    3   

     total_protein outcome  
232            7.7       1  
130            NaN       2  
276           75.0       3  
203            5.7       1  
76            68.0       1  
..             ...     ...  
63            82.0       3  
23            70.0       1  
148            8.0       1  
135            4.9       2  
103            5.3       1  

[214 rows x 8 columns]

In [150]:
new_df = find_outlier(new_df, 'total_protein')
new_df

surgery? Age Hospital_Number  rectal_temperature  pulse pain  \
210        1   1          530401                37.9   68.0    2   
84         1   1          529849                37.8   60.0    2   
247        1   1          530170                38.1   88.0    5   
135        1   9         5287179                38.1  136.0    5   
287        2   9         5305129                39.5   84.0  NaN   
..       ...  ..             ...                 ...    ...  ...   
14         1   1          534917                38.2   76.0    3   
63         2   1          534293                38.2  130.0    2   
1          1   1          534817                39.2   88.0    3   
90         2   1          534197                38.0   52.0    2   
208        1   1          535338                37.8   88.0  NaN   

     total_protein outcome  
210            4.0       2  
84             4.5       1  
247            4.6       2  
135            4.9       2  
287            5.0       1  
..             ...     ...  
14            81.0       1  
63            82.0       3  
1             85.0       3  
90            86.0       1  
208           89.0       3  

[196 rows x 8 columns]

Получили датафрейм без строк, содержащих выбросы.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [154]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 196 entries, 210 to 208
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   surgery?            195 non-null    object 
 1   Age                 196 non-null    object 
 2   Hospital_Number     196 non-null    object 
 3   rectal_temperature  196 non-null    float64
 4   pulse               196 non-null    float64
 5   pain                166 non-null    object 
 6   total_protein       196 non-null    float64
 7   outcome             195 non-null    object 
dtypes: float64(3), object(5)
memory usage: 13.8+ KB


После предыдущего отсеивания у нас осталось 3 столбца с пропусками. Это: surgery?, pain и outcome. В случае с surgery? и outcome: и там и там присутствует по одному пропуску. Это важные столбцы, без которых статистика теряет смысл, соответственно строки с пропусками в этих столбцах следует удалить. А вот в  случае с pain всё неоднозначно. В зависимости от того, что нам предстоит исследовать, строки с пропусками в этом столбце можно как удалять, так и игнорировать. Например, если мы будем исследовать зависимость пульса от испытываемой боли у лошади, то имеет смысл строки удалить. Если же мы будем проверять какую-то гепотезу, в которую не включен столбец pain, либо будем исследовать столбец pain обособленно, то пропуски можно игнорировать. Однозначно можно сказать только одно: нет смысла самим заполнять пропуски в этом столбце (если только мы не найдём связь столбца pain с типом операции или чем-то другим, что позволит нам эти стобцы заполнить).  

Т.к. по заданию нужно сформировать датафрейм, в котором пропуски будут отсутствовать, то имеет смысл просто удалить стобец pain.

In [158]:
new_df = new_df.drop(['pain'], axis = 1)
new_df

surgery? Age Hospital_Number  rectal_temperature  pulse  total_protein  \
210        1   1          530401                37.9   68.0            4.0   
84         1   1          529849                37.8   60.0            4.5   
247        1   1          530170                38.1   88.0            4.6   
135        1   9         5287179                38.1  136.0            4.9   
287        2   9         5305129                39.5   84.0            5.0   
..       ...  ..             ...                 ...    ...            ...   
14         1   1          534917                38.2   76.0           81.0   
63         2   1          534293                38.2  130.0           82.0   
1          1   1          534817                39.2   88.0           85.0   
90         2   1          534197                38.0   52.0           86.0   
208        1   1          535338                37.8   88.0           89.0   

    outcome  
210       2  
84        1  
247       2  
135       2  
287       1  
..      ...  
14        1  
63        3  
1         3  
90        1  
208       3  

[196 rows x 7 columns]

In [161]:
new_df = new_df.dropna()
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 195 entries, 210 to 208
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   surgery?            195 non-null    object 
 1   Age                 195 non-null    object 
 2   Hospital_Number     195 non-null    object 
 3   rectal_temperature  195 non-null    float64
 4   pulse               195 non-null    float64
 5   total_protein       195 non-null    float64
 6   outcome             195 non-null    object 
dtypes: float64(3), object(4)
memory usage: 12.2+ KB
